vowpalwabbit/learner.h

```diff
@@ -465,9 +515,51 @@ VW_WARNING_STATE_POP
     ret.learn_fd.multipredict_f = nullptr;
     ret.pred_type = pred_type;
     ret.is_multiline = std::is_same<multi_ex, E>::value;
+    ret._hash_index = std::type_index(typeid(T));
 
     return ret;
   }
+
+  std::type_index hash_index() { return _hash_index; }
+
+  inline VW::LEARNER::base_learner* get_base_reduction() { return learn_fd.base; }
+
+  void apply_from(
+    const VW::LEARNER::base_learner* base,
+    const std::unordered_map<std::type_index, VW::LEARNER::base_learner*>& reduction_template_map) {
+    if (base == nullptr || get_base_reduction() == nullptr) {
+      // currently working on base type. Nothing to do
+      return;
+    }
+    auto src = *this;
+    *this = *base;
+
+    _hash_index = src._hash_index;
+    auto src_template_it = reduction_template_map.find(src._hash_index);
+    if (src_template_it == reduction_template_map.end()) THROW("invalid template map");
+    auto * src_template = src_template_it->second;
+    auto * noop_template = src_template->learn_fd.base;
+
+    if (noop_template->init_fd != src_template->init_fd) init_fd = src.init_fd;
+    if (noop_template->sensitivity_fd != src_template->sensitivity_fd) sensitivity_fd = src.sensitivity_fd;
+    if (noop_template->finish_example_fd != src_template->finish_example_fd) finish_example_fd = src.finish_example_fd;
+    if (noop_template->save_load_fd != src_template->save_load_fd) save_load_fd = src.save_load_fd;
+    if (noop_template->end_pass_fd != src_template->end_pass_fd) end_pass_fd = src.end_pass_fd;
+    if (noop_template->end_examples_fd != src_template->end_examples_fd) end_examples_fd = src.end_examples_fd;
+    if (noop_template->pred_type != src_template->pred_type) pred_type = src.pred_type;
+    weights = src.weights;
+    increment = base->increment * src.weights;
+
+    _hash_index = src._hash_index;
```

Comment:
are these the things that change when some _setup functions get called?

Reply:
This function does the following:
1. copies the base learner
1. overwrites everything that is set in `setup_base`
1. overwrites anything that was modified in the setup function (on the learner only)